public/js/reducers/sources.js

```diff
@@ -167,41 +166,32 @@ function updateTabList(state, source, tabIndex) {
 }
 
 /**
- * Gets the next tab to select when a tab closes.
+ * Gets the next tab to select when a tab closes. Heuristics:
+ * 1. if the selected tab is available, it remains selected
+ * 2. if it is gone, the next available tab to the left should be active
+ * 3. if the first tab is active and closed, select the second tab
+ *
  * @memberof reducers/sources
  * @static
  */
-function getNewSelectedSourceId(state, id: any) : ?Source {
-  // If id is an array
-  const isTabSet = typeof id !== "string";
-
-  const tabs = state.get("tabs");
-  const numTabs = tabs.count();
-  const numClosedTabs = !isTabSet ? 1 : id.count();
-
-  if (numTabs == numClosedTabs) {
-    // If all the tabs are closed
+function getNewSelectedSourceId(state, availableTabs: List<Source>) : ?Source {
+  if (!state.selectedLocation) {
     return undefined;
   }
 
-  if (isTabSet) {
-    // if 1 or more tabs are closed but not all
-    let selectTabs = tabs;
-    id.forEach(i => {
-      selectTabs = selectTabs.filter(tab => tab.get("id") !== i);
-    });
-    return selectTabs.get(selectTabs.count() - 1).get("id");
-  }
+  const selectedTabId = state.selectedLocation.sourceId;
+  const availableTabIds = availableTabs.map(s => s.get("id")).toJS();
+  const tabIds = state.tabs.map(s => s.get("id")).toJS();
```

Comment:
it transforms an immutable record, in this case list to a native type, in this case array. Makes things simpler :)